### PR TITLE
[FEATURE] Validation Reports API

### PR DESCRIPTION
**Pull request by stevenzeiler**

### CODE REVIEW

File: api/services/ValidationReportService.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+
+export async function historyForValidator(validationPublicKey) {
+
+  const reports = await ValidationReport.find()
+
+  const relevantReports = _.reject(reports, report => {
+    return _.has(report, validationPublicKey)
```
Comment (stevenzeiler) [inline api/services/ValidationReportService.js]: should be `!_has`

Comment (stevenzeiler) [inline api/services/ValidationReportService.js]: actually, should be `_.filter` not `_.reject`

File: api/services/ValidationReportService.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    }
+  })
+}
+
+export async function latest() {
+
+  const reports = await ValidationReport.find({ order: 'date DESC', limit: 1 })
+
+  return {
+    date: '1605-11-05',
```
Comment (wilsonianb) [inline api/services/ValidationReportService.js]: 1605!

File: api/services/ValidationReportService.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      validators: validators
+    })
+    return report
+  }
+}
+
+export async function compute(date) {
+
+  const start = moment(date).startOf('day').format(DATE_FORMAT)
+  const end   = moment(date).startOf('day').add(1, 'day').format(DATE_FORMAT)
```
Comment (wilsonianb) [inline api/services/ValidationReportService.js]: is `startof('day')` necessary?

File: api/services/ValidationReportService.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import moment from 'moment'
+
+const DATE_FORMAT = 'YYYY-MM-DD'
+
+export async function historyForValidator(validationPublicKey) {
+
+  const reports = await database.ValidationReports.findAll()
+
+  const relevantReports = _.reject(reports, report => {
+    return _.has(report, validationPublicKey)
```
Comment (wilsonianb) [inline api/services/ValidationReportService.js]: `report` -> `report.validators`?

File: api/services/ValidationReportService.js (hunk 3 above)
Comment (stevenzeiler) [inline api/services/ValidationReportService.js]: probably not

File: tasks/db_seed.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  const publicKeys = [
+    'n9KHeNRQCgmzdD2WF92abMkRfxWGFyYQS1rJ5idZt9Yg3fEXU6Dh',
+    'n9JmgHszbQFCm6zD8NSobkd7K54qLtDgCaYf2mq772N5KMYEmnXv',
+    'n9L81uNCaPgtUJfaHh89gmdvXKAmSt5Gdsw2g1iPWaPkAHW5Nm4C'
+  ]
+
+  const day1 = moment().subtract(1, 'day')
+  const day2 = moment().subtract(2, 'days')
+  const day3 = moment().subtract(3, 'days')
```
Comment (wilsonianb) [inline tasks/db_seed.js]: What do you think about only dealing with dates with `.format('YYYY-MM-DD')` ?

Comment (wilsonianb) [inline tasks/db_seed.js]: Ah, I see that we need the more specific datetime for the `createdAt` below.
Then how about having `ValidationReportService.create()` only take dates with `.format('YYYY-MM-DD')`?

Comment (stevenzeiler) [inline tasks/db_seed.js]: `createdAt` is automatically set by sequelize to the exact timestamp when the record was created. Think of it more as record metadata.

File: tasks/db_seed.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    for (let k=0; k<publicKeys.length; k++) {
+
+      for (let i=0; i<5; i++) {
+
+        try {
+          let v = await database.Validations.create({
+            validation_public_key: publicKeys[k],
+            reporter_public_key: publicKeys[k],
+            ledger_hash: SHA256(),
+            createdAt: days[j]
```
Comment (wilsonianb) [inline tasks/db_seed.js]: Remove this if we're setting it with `.updateAttribute` below

File: test/controllers/ValidationReportsControllerTest.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    let validation = await database.Validations.create({
+      validation_public_key: publicKey,
+      reporter_public_key: publicKey,
+      ledger_hash: SHA256(),
+      createdAt: day.toDate()
+    })
+    await ValidationReportService.create(day)
+
+    request.get('/reports').end((error, response) => {
```
Comment (wilsonianb) [inline test/controllers/ValidationReportsControllerTest.js]: Add `.expect(200)` here and below

File: test/controllers/ValidationReportsControllerTest.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      assert(response.body.report.date)
+      assert(response.body.report.validators)
+      done()
+    })
+  })
+
+  it('#show should return the historical daily report for a validadtor', (done) => {
```
Comment (wilsonianb) [inline test/controllers/ValidationReportsControllerTest.js]: `validadtor` -> `validator`

File: api/models/validation.js [hunk 9]
```diff
@@ -27,6 +27,9 @@ module.exports = function(sequelize, DataTypes) {
       type     : DataTypes.STRING,
       allowNull: false
     },
+    createdAt: {
+      type     : DataTypes.DATE
```
Comment (wilsonianb) [inline api/models/validation.js]: `allowNull: false` ?

File: api/models/ValidationReport.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+          msg: 'date must be formatted as YYYY-MM-DD'
+        }
+      }
+    }
+  }, {
+    getterMethods: {
+      validators: function() {
+        var validators = this.getDataValue('validators')
+        Object.keys(validators).forEach(publicKey => {
+          validators[publicKey] = parseFloat(validators[publicKey])
```
Comment (wilsonianb) [inline api/models/ValidationReport.js]: This is a float?

File: api/models/validation.js (hunk 9 above)
Comment (stevenzeiler) [inline api/models/validation.js]: Sequelize automatically populates `createdAt`, so we should allowNull: true to allow sequelize to do its auto-populate job.

File: tasks/db_seed.js (hunk 6 above)
Comment (stevenzeiler) [inline tasks/db_seed.js]: Yeah, plus it isn't possible to set `createdAt` anyway I discovered.

File: tasks/db_seed.js (hunk 5 above)
Comment (stevenzeiler) [inline tasks/db_seed.js]: We could have `ValidationReportService.create()` take only formatted dates, or date objects I am ambivalent.

File: api/models/ValidationReport.js (hunk 10 above)
Comment (stevenzeiler) [inline api/models/ValidationReport.js]: hmmm no it should be `parseInt`

File: tasks/db_seed.js (hunk 5 above)
Comment (wilsonianb) [inline tasks/db_seed.js]: We were sticking to formatted dates in `CorrelationCoefficientService`. It was save us from having to convert dates in `ValidationReportService`

Comment (stevenzeiler) [inline tasks/db_seed.js]: Do you want to make that update?

Comment (wilsonianb) [inline tasks/db_seed.js]: sure

File: test/services/ValidationReportServiceTest.js [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    const day1 = moment().subtract(1, 'day')
+    const day2 = moment().subtract(2, 'days')
+    const day3 = moment().subtract(3, 'days')
+
+    for (let i=0; i<3; i++) {
+      let validation = await database.Validations.create({
+        validation_public_key: publicKey1,
+        reporter_public_key: publicKey1,
+        ledger_hash: SHA256(),
+        createdAt: day1.toDate()
```
Comment (wilsonianb) [inline test/services/ValidationReportServiceTest.js]: Don't set `createdAt` here or below

Comment (stevenzeiler) [inline test/services/ValidationReportServiceTest.js]: Apparently I need to set createdAt here in order to get the fake seed data I want for development purposes.

**Comment by stevenzeiler** [comment]
:+1:

File: tasks/db_seed.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import Sails from 'sails'
+import moment from 'moment'
+import bluebird from 'bluebird'
```
Comment (wilsonianb) [inline tasks/db_seed.js]: `bluebird` isn't listed as a dependency in package.json

**Comment by wilsonianb** [comment]
LGTM :+1:

**Comment by wilsonianb** [comment]
Slack test

**Pull request merged**